rem/rem.man
.\"_
.TH REM 1 "1981-02-23" PCS
.SH NAME
rem \- request a reminder notification
.SH SYNOPSIS
.\"_
.so tmac.att
.\"_
.B rem
hour:minute 'message'
.br
.B rem
-
.br
.B rem
\(+-mail
.br
.B rem
+today

.br
.B remday
day hour:minute 'message'
.\"_
.ad b
.PM BP
.\"_
.SH DESCRIPTION
.I Rem
logs requests for reminder notifications.
Reminder items consist of a time and a message.
The time is of the form 
"hours:minutes" (e.g. 9:15) or "hours" (e.g. 20).
The hour may be specified on a 24 hour clock.
Alternatively, hours specified on a 12 hour clock are assumed to 
be between 8 AM and 7 PM.
This is done by converting times between 1:00 and 7:59 to the corresponding
24 hour time (i.e. 1:00 to 7:59 in the afternoon).
.P
Every five minutes, the reminder demon checks whether any reminder messages 
should be announced.  If so, it writes the message and the current time to
the user's terminal.
A message is announced when it is near to its specified time.
The definition of "near" is given in the environment variable,
\f3REM_TIMES\f1, set by the PCS profiler.  The default value of this
is 30, 10, and 5 minutes.  Thus the reminder is announced 30 minutes
before, 10 minutes before, 5 minutes before, and at the specified time
of the message.
The values of these warning variables may range between 0 and 90 minutes;
if 0 no announcement is given.
.P  
If 
.I rem 
is invoked with the "-" argument, reminder records will be read from
the standard input.  Each record consists of one line containing
"hour:minutes\ message".
.P
The +mail and -mail options respectively add and delete a user 
from the mail notification list.
If a user is in the mail notification list, that user's mailbox will be checked
every five minutes by the reminder service demon to determine
if mail has arrived within that period.
If so, the message "new mail arrived" will be issued
to all terminals on which the user is currently active (logged in).
Once added to the notification list, a user remains on it until a
"rem\ -mail" command is explicitly entered.
.P
The +today option displays all of a user's reminders
for the current day. These will include items passed
to the reminder service by "cal2rem" during the
standard PCS startup.
.P
Reminders are for the current day, since all reminder items enqueued are removed at midnight.
The cal2rem program, invoked automatically by the 
default PCS startup, places calendar items for today into the reminder
queue.
.I Rem 
places each reminder in a unique file in the reminder directory.
These items are merged into the reminder file by the reminder monitor after
its next five minute pass through the file.
Reminder items, once entered, may not be removed.
The REM_TIMES variable is used when rem enqueues the reminder, not dynamically
during the day.
.I Remday
accepts requests for reminder notifications.
If the day specified is the current day of the week,
the reminder is passed to 
.I rem
in the usual fashion; otherwise
.I remday
has no effect.
.I Remday
is designed to be placed in a user's .profile file to automatically
request reminders for specific days of the week.
For example:
.EX
remday fri 8:45 fill out time card
.EE
The day may be specified in as few letters as necessary to distinguish
it from another day. (i.e. fr, tue, etc.)

.SH FILES
.PD 0
.TP 30
${PCSDIR}/src/rem
directory of source files
.TP
${PCSBIN}/rem
executable program
.TP
${PCSBIN}/cal2rem
executable program
.TP
${PCSBIN}/remday
executable program
.TP
${PCSLIB}/remdata/remfile
master reminder queue
.TP
${PCSLIB}/remdata/rem.*
items not yet queued
.TP
${PCSLIB}/remdata/remtoday
executable program
.TP
${PCSLIB}/remdata/mc_users
users desiring mail notification
.PD
.SH "SEE ALSO"
profiler (PCS)
.\"_


rshe/CHANGES
RSHE



= Dave Morano, 95/08/22
version '0a'

This program was originally written.  It was inspired by the
'bopenrcmd' as part of the "Basic I/O" (BIO) package.


= Dave Morano, 95/10/12
version '0b'

1) The meaning of the 'x' program invocation option was changed
   to match other such modern "remote execution" programs such
   as 'rex'.





pcsrem/rem.c
/* rem: places items in reminder service directory.
* J. A. Kutsch  ho 43233   x-3059
* July 1980
* converted from extended shell script to c in January 1981
*/

#define REMDIR "/usr/add-on/pcs/lib/remdata"

#include "remfile.h"
#include <stdio.h>
#include <string.h>
struct msgrec msgrec;
FILE *outfile, *fopen();
main(argc,argv)
int argc;
char **argv;
{
    int i;
    char tempfile[50], realfile[50];
    char buffer[150];
    char hrbuf[2], minbuf[2],msgbuf[150];
    char unique[16];
    char s[256];
    char *edata, *getenv();

    char *adata;	/*string for three times in one */
    char *tok;

    if (strcmp(argv[1],"+mail")==0) {
        strcpy(tempfile,REMDIR);
        strcat(tempfile,"/mailon");
        system (tempfile);
        exit(0);
    }

    if (strcmp(argv[1],"-mail")==0) {
        strcpy(tempfile,REMDIR);
        strcat(tempfile,"/mailoff");
        system (tempfile);
        exit(0);
    }

    if ((strcmp(argv[1],"+today"))==0 || (strcmp(argv[1],"+t"))==0 || (strcmp(argv[1],"+to"))==0){
        strcpy(tempfile,REMDIR);
        strcat(tempfile,"/remtoday");
        system(tempfile);
        exit(0);
    }
    edata=getenv("LOGNAME");
    if (edata==NULL)
        fatal("LOGNAME not set in environment");
    strcpy(msgrec.name,edata);
    for (i=strlen(msgrec.name);i<sizeof(msgrec.name);i++)
        msgrec.name[i]=' ';

    /* Try for REM_TIMES */
    if((adata=getenv("REM_TIMES")) == NULL)
    {
        edata=getenv("REM_T1");
        if (edata==NULL)
            msgrec.t1[0]=msgrec.t1[1]='0';
        else tdata(edata,msgrec.t1);	/* convert to zero fill, right justified */
        edata=getenv("REM_T2");
        if (edata==NULL)
            msgrec.t2[0]=msgrec.t2[1]='0';
        else tdata(edata,msgrec.t2);	/* convert to zero fill, right justified */
        edata=getenv("REM_T3");
        if (edata==NULL)
            msgrec.t3[0]=msgrec.t3[1]='0';
        else tdata(edata,msgrec.t3);	/* convert to zero fill, right justified */
    }

    else
    {
        sprintf (s,adata);
        tok = strtok (s,":");
        if (tok != NULL) tdata (tok, msgrec.t1);
        tok = strtok (0,":");
        if (tok !=NULL) tdata (tok, msgrec.t2);
        tok = strtok (0,":");
        tdata (tok,msgrec.t3);
    }
    strcpy(unique,"XXXXXX");
    mktemp(unique);
    strcpy(tempfile,REMDIR);
    strcat(tempfile,"/TMP.");
    strcat(tempfile,unique);
    strcpy(realfile,REMDIR);
    strcat(realfile,"/rem.");
    strcat(realfile,unique);
    umask(0);  /* make created files readable */
    if ((outfile=fopen(tempfile,"w")) == NULL)
        fatal("cannot create tempfile");
    if (strcmp(argv[1],"-")!=0) {
        buffer[0]='\0';
        for (i=1;i<argc;i++) {
            strcat(buffer," ");
            strcat(buffer,argv[i]);
        }
        if (parse(buffer,hrbuf,minbuf,msgbuf))
            insert(hrbuf,minbuf,msgbuf);
        else fprintf(stderr,"Rem: bad time field\n");
    }
    else /* read from stdin */
        while (gets(buffer)!=NULL) {
            if (parse(buffer,hrbuf,minbuf,msgbuf))
                insert(hrbuf,minbuf,msgbuf);
        }

    fclose(outfile);
    sync();
    link(tempfile,realfile);
    unlink(tempfile);
}

insert (hourarg,minarg,msgarg)
char *hourarg, *minarg, *msgarg;
{
    char *digits = "0123456789";
    msgrec.hour[0]=hourarg[0];
    msgrec.hour[1]=hourarg[1];
    msgrec.min[0]=minarg[0];
    msgrec.min[1]=minarg[1];
    if (strncmp(msgrec.hour,"08",2)<0) {
        /* times before 8 a.m. are assumed to be 1200 through 1959 */
        msgrec.hour[0]='1';
        msgrec.hour[1]=digits[instr(digits,msgrec.hour[1])+2];
    }

    strcpy(msgrec.text,msgarg);
    fprintf(outfile,"%s\n",&msgrec);
}

fatal(msg)
char *msg;
{
    fprintf(stderr,"rem: %s\n",msg);
    exit(1);
}

instr(s,c)
/* uses index to see where char c is in string s */
char *s, c;
{
    char s2[2];
    s2[0]=c;
    s2[1]='\0';
    return(index(s,s2));
}
